src/sql/src/plan/statement/ddl/connection.rs

```diff
@@ -164,17 +169,102 @@ impl ConnectionOptionExtracted {
+                                ))
+                            })?;
+                        Some(AwsAuth::AssumeRole(AwsAssumeRole {
+                            arn,
+                            session_name,
+                            mz_principal: MaterializePrincipal {
+                                arn: mz_arn,
+                                external_id_prefix,
+                            },
+                        }))
+                    }
+                    (None, Some(_)) => {
+                        return Err(PlanError::MissingRequiredOptions {
+                            option_names: vec![ConnectionOptionName::AssumeRoleArn.to_string()],
+                            item_type: CatalogItemType::Connection,
+                            item_sub_type: Some("AWS AssumeRole".to_string()),
+                        })
+                    }
+                    _ => None,
+                };
+
+                if credentials.is_some() && assume_role.is_some() {
+                    return Err(PlanError::ConflictingOptions {
+                        option_names: vec![
+                            format!(
+                                "{} and {}",
+                                ConnectionOptionName::AccessKeyId,
+                                ConnectionOptionName::SecretAccessKey
+                            ),
+                            ConnectionOptionName::AssumeRoleArn.to_string(),
+                        ],
+                        item_type: CatalogItemType::Connection,
+                        item_sub_type: Some(CreateConnectionType::Aws.to_string()),
+                    });
+                }
+
                 Connection::Aws(AwsConfig {
-                    credentials: AwsCredentials {
-                        access_key_id: self
-                            .access_key_id
-                            .ok_or_else(|| sql_err!("ACCESS KEY ID option is required"))?,
-                        secret_access_key: self
-                            .secret_access_key
-                            .ok_or_else(|| sql_err!("SECRET ACCESS KEY option is required"))?
-                            .into(),
-                        session_token: self.token,
-                    },
+                    auth: credentials.or(assume_role).ok_or_else(|| {
+                        PlanError::MissingRequiredOptions {
+                            option_names: vec![format!(
+                                "{} and {} or {}",
+                                ConnectionOptionName::AccessKeyId,
+                                ConnectionOptionName::SecretAccessKey,
+                                ConnectionOptionName::AssumeRoleArn
+                            )],
+                            item_type: CatalogItemType::Connection,
+                            item_sub_type: Some(CreateConnectionType::Aws.to_string()),
+                        }
+                    })?,
```

Comment:
This is what I actually had before https://github.com/MaterializeInc/materialize/pull/23282#discussion_r1414155766

Reply:
I think I will revert back to using `sql_bail!` here. From the feedback it's making readability worse.

cc @jkosh44 @sploiselle Do you have better ideas here?

Reply:
I really appreciate how you threaded the needle in making sure all of the possible states are covered here. One thing that might make the code a little more verbose would be to make the errors slightly less context sensitive––for example, you could error with all of the required options if any of them are missing. Users can do a little bit of deduction to figure out what they need.

Another thing that might help is to use the concrete types in these new error types and then rely on converting them to strings when displaying the error––this would get rid of the `to_string` noise, as well as coerce you to pare the errors down to only the minimal amount of context.

Glad to chat through any of this!

Reply:
> you could error with all of the required options if any of them are missing.

That's a good point. Let me do that, should clean up a bunch of match arms.